presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -141,14 +141,7 @@ public SqlStageExecution(StageStateMachine stateMachine, RemoteTaskFactory remot
         }
         this.exchangeSources = fragmentToExchangeSource.build();
 
-        stateMachine.addStateChangeListener(newState -> {
-            synchronized (this) {
-                // when stage transitions to a done state, check if all tasks have final status information
-                if (newState.isDone() && doneTasks.containsAll(allTasks)) {
-                    stateMachine.setAllTasksFinal();
-                }
-            }
-        });
+        stateMachine.addStateChangeListener(newState -> checkAllTaskFinal());
```

Comment:
Do we need to check for finalTaskInfo here?

Reply:
`checkAllTaskFinal`  does that check

Reply:
No, i mean if we still need this check here. I think we do, for a case when no tasks were scheduled.